### PR TITLE
fix(ses): Defend integrity of intrinsics

### PR DESCRIPTION
**Pull request by kriskowal**
In order for the SES shim to faithfully approximate the intended behavior of SES as implemented by engines, it must capture all of the intrinsics it uses and avoid dynamic dispatch on any object that might be modified after SES initialization. After lockdown, the shared intrinsics are immutable, but all execution between initialization and lockdown in the privileged start compartment could alter the intrinsics and compromise the correctness of SES. Also, SES may depend upon free but unshared intrinsics (though I can think of no example).

This is not a security vulnerability because SES provides no guarantees against interference from code running in the start compartment, only code running in child compartments. This change does address a weakness in the SES shim’s integrity against accidental mutation to the shared platform.

### CODE REVIEW

File: packages/ses/src/module-instance.js [hunk 1]
```diff
@@ -1,8 +1,20 @@
...
+  TypeError,
+  create,
+  defineProperty,
+  entries,
+  freeze,
+  isArray,
+  keys,
+} from './commons.js';
+
+const { quote: q } = assert;
```
Comment (erights) [inline packages/ses/src/module-instance.js:17]: This is an improvement, but is not a mechanical-ish consequence of the pattern used in the rest of this PR. Makes me wonder, what other semantic changes are there in this PR I may miss when reading only for fidelity to this PR's purpose? Please mention all such semantic changes in the PR's comment.

File: packages/ses/src/commons.js [hunk 2]
```diff
@@ -135,4 +199,34 @@ export const getConstructorOf = fn =>
  * immutableObject
  * An immutable (frozen) exotic object and is safe to share.
  */
-export const immutableObject = freeze({ __proto__: null });
+export const immutableObject = freeze(create(null));
```
Comment (kriskowal) [inline packages/ses/src/commons.js:202]: This is a rider bill.

File: packages/ses/src/commons.js [hunk 3]
```diff
@@ -135,4 +199,34 @@ export const getConstructorOf = fn =>
...
+ *   };
+ *
+ * But this is not safe in the face of possible evolution of the language, for
+ * example new types or semantics of records and tuples.
+ * We use this implementation despite the unnecessary allocation implied by
+ * attempting to box a primitive.
+ *
+ * @param {any} value
+ */
+export const isObject = value => Object(value) === value;
```
Comment (kriskowal) [inline packages/ses/src/commons.js:221]: More pork. Not all uses of `isObject` were previously defined this way, but as we discussed out-of-band, they all should work with this generalization, which favors integrity in the face of language evolution over memory pressure performance.

File: packages/ses/src/enable-property-overrides.js [hunk 4]
```diff
@@ -17,12 +24,6 @@ import {
   severeEnablements,
 } from './enablements.js';
 
-const { ownKeys } = Reflect;
-
-function isObject(obj) {
-  return obj !== null && typeof obj === 'object';
-}
```
Comment (kriskowal) [inline packages/ses/src/enable-property-overrides.js:24]: Tangential to the primary thrust.

File: packages/ses/src/get-anonymous-intrinsics.js [hunk 5]
```diff
@@ -26,10 +38,7 @@ function makeArguments() {
  * @returns {Object}
  */
 export const getAnonymousIntrinsics = () => {
-  const InertFunction = Function.prototype.constructor;
-
-  const SymbolIterator = (typeof Symbol && Symbol.iterator) || '@@iterator';
-  const SymbolMatchAll = (typeof Symbol && Symbol.matchAll) || '@@matchAll';
```
Comment (kriskowal) [inline packages/ses/src/get-anonymous-intrinsics.js:32]: Drive-by fix. These lines were defective, not testing the value of each typeof expression. Beyond removing the defective code, using `iteratorSymbol` and `matchAllSymbol` from `commons.js` is also spiritually different than the original intended semantics. As a matter of fact, we never supported versions before `Symbol` because of this defect. As a matter of stated supported versions, we no longer intend to.

Fixes #828

File: packages/ses/src/intrinsics.js [hunk 6]
```diff
@@ -17,6 +21,8 @@ import {
   whitelist,
 } from './whitelist.js';
 
+const isFunction = obj => typeof obj === 'function';
```
Comment (kriskowal) [inline packages/ses/src/intrinsics.js:24]: This participates in an incidental refactor below, avoiding an unnecessary closure allocation for filtering functions.

File: packages/ses/src/lockdown-shim.js [hunk 7]
```diff
@@ -184,13 +183,13 @@ export const repairIntrinsics = (
...
+      globalThis.Function.prototype.constructor !== globalThis.Function &&
       typeof globalThis.harden === 'function' &&
       typeof globalThis.lockdown === 'function' &&
-      Date.prototype.constructor !== Date &&
-      typeof Date.now === 'function' &&
+      globalThis.Date.prototype.constructor !== globalThis.Date &&
+      typeof globalThis.Date.now === 'function' &&
       // @ts-ignore
-      Object.is(Date.prototype.constructor.now(), NaN)
+      is(globalThis.Date.prototype.constructor.now(), NaN)
```
Comment (kriskowal) [inline packages/ses/src/lockdown-shim.js:192]: These could probably be framed in terms of captured intrinsics, but I instead used `globalThis` pervasively to avoid changing the late-binding semantics.

File: packages/ses/src/make-hardener.js [hunk 8]
```diff
@@ -53,7 +72,7 @@ export const makeHardener = () => {
        * @param {string} [path]
        */
       function enqueue(val, path = undefined) {
-        if (Object(val) !== val) {
+        if (!isObject(val)) {
```
Comment (kriskowal) [inline packages/ses/src/make-hardener.js:75]: Not quite mechanical, changing an inlined behavior to a commons function.

File: packages/ses/src/make-hardener.js [hunk 9]
```diff
@@ -88,27 +107,25 @@ export const makeHardener = () => {
...
           // test could be confused. We use hasOwnProperty to be sure about
           // whether 'value' is present or not, which tells us for sure that
           // this is a data property.
-          // The 'name' may be a symbol, and TypeScript doesn't like us to
-          // index arbitrary symbols on objects, so we pretend they're just
-          // strings.
-          const desc = descs[/** @type {string} */ (name)];
-          if ('value' in desc) {
-            // todo uncurried form
+          if (objectHasOwnProperty(desc, 'value')) {
```
Comment (kriskowal) [inline packages/ses/src/make-hardener.js:128]: A non-mechanical detail here is reärrangement of the comments to clarify their respective subjects. This was a point of confusion in @dominictarr’s review.

File: packages/ses/src/module-link.js [hunk 10]
```diff
@@ -13,14 +13,21 @@ import {
...
+  mapHas,
+  mapSet,
+  weakmapGet,
+} from './commons.js';
 
 const { quote: q } = assert;
 
-function isObject(o) {
-  return Object(o) === o;
-}
```
Comment (kriskowal) [inline packages/ses/src/module-link.js:22]: Incidental refactor to reuse a shared implementation. In this case, the shared version is identical.

File: packages/ses/src/module-proxy.js [hunk 11]
```diff
@@ -11,10 +11,21 @@
...
+  reflectGetOwnPropertyDescriptor,
+  reflectHas,
+  reflectIsExtensible,
+  reflectPreventExtensions,
+} from './commons.js';
+import { assert } from './error/assert.js';
 
-// q, as in quote, for error messages.
-const q = JSON.stringify;
+const { quote: q } = assert;
```
Comment (kriskowal) [inline packages/ses/src/module-proxy.js:29]: Another drive-by change. The mechanical change would have been to use `stringifyJson` from `commons.js`, but I’ve been meaning to refactor the lines of `const q = JSON.stringify` to use the new assert function anyway.

File: packages/ses/src/module-proxy.js [hunk 12]
```diff
@@ -79,8 +90,7 @@ export const deferExports = () => {
             'Cannot enumerate keys, the module has not yet begun to execute',
           );
         }
-        // return Object.keys(proxiedExports);
```
Comment (kriskowal) [inline packages/ses/src/module-proxy.js:82]: I don’t know why the commented out line above existed. It might need a “beware of dog” sign.

File: packages/ses/src/scope-constants.js [hunk 13]
```diff
@@ -84,7 +84,7 @@ const keywords = [
  * Note: \w is equivalent [a-zA-Z_0-9]
  * See 11.6.1 Identifier Names
  */
-const identifierPattern = new RegExp('^[a-zA-Z_$][\\w$]*$');
+const identifierPattern = /^[a-zA-Z_$][\w$]*$/;
```
Comment (kriskowal) [inline packages/ses/src/scope-constants.js:87]: Instead of taking a dependency on `RegExp` from `commons.js`, I folded an unrelated recommended change to use a regex literal here.

File: packages/ses/src/scope-handler.js [hunk 14]
```diff
@@ -10,12 +13,6 @@ import { assert } from './error/assert.js';
 
 const { details: d, quote: q } = assert;
 
-// The original unsafe untamed eval function, which must not escape.
-// Sample at module initialization time, which is before lockdown can
-// repair it.  Use it only to build powerless abstractions.
-// eslint-disable-next-line no-eval
-const FERAL_EVAL = eval;
```
Comment (kriskowal) [inline packages/ses/src/scope-handler.js:17]: I’ll note that there were duplicates of this block, complete with matching comments. They are now in `commons.js`.

File: packages/ses/src/scope-handler.js [hunk 15]
```diff
@@ -162,7 +159,7 @@ export const createScopeHandler = (
 
     getOwnPropertyDescriptor(_target, prop) {
       // Coerce with `String` in case prop is a symbol.
-      const quotedProp = JSON.stringify(String(prop));
+      const quotedProp = q(String(prop));
```
Comment (kriskowal) [inline packages/ses/src/scope-handler.js:162]: This is another incidental refactor for quoting. I don’t know whether we need the inner `String` for coercing symbols anymore.

File: packages/ses/src/whitelist-intrinsics.js [hunk 16]
```diff
@@ -89,7 +93,7 @@ export default function whitelistIntrinsics(
    * Validate the object's [[prototype]] against a permit.
    */
   function whitelistPrototype(path, obj, protoName) {
-    if (obj !== Object(obj)) {
```
Comment (kriskowal) [inline packages/ses/src/whitelist-intrinsics.js:92]: More pork. This is no longer inlined.

File: packages/ses/src/module-instance.js (hunk 1 above)
Comment (kriskowal) [reply packages/ses/src/module-instance.js:17]: Added a supplementary review with annotations for the non-mechanical changes.

**Comment by kriskowal** [comment]
Appended a commit to propagate a change necessitated by an optimization in the prior change on the stack #824

File: packages/ses/src/scope-handler.js (hunk 15 above)
Comment (erights) [reply packages/ses/src/scope-handler.js:162]: See https://github.com/endojs/endo/blob/f07a8b3fa670a219e06f1a7ba04524d50de85a1e/packages/ses/src/error/stringify-utils.js#L101
It is not quite equivalent. But if you agree that it is adequately equivalent, then
```suggestion
      const quotedProp = q(prop);
```

**Review by erights — APPROVED** [review]
LGTM!
